### PR TITLE
Feat: Adding file through Notion and onedrive

### PR DESCRIPTION
**Pull request by StrongMonkey**
This PR adds ability to add files through notion and onedrive when using knowledge.

It mainly use https://github.com/gptscript-ai/knowledge-notion-integration and https://github.com/gptscript-ai/knowledge-onedrive-integration to sync down files from notion/onedrive to local disk and keep those up-to-sync. The sync will be run every 24 hours or can be triggered manually.

https://github.com/gptscript-ai/desktop/issues/323

### CODE REVIEW

File: components/edit/configure.tsx
Comment (ryanhopperlowe) [inline components/edit/configure.tsx:232]: nit: might be easier to read if you destructure the entries:
```suggestion
                  {Array.from(droppedFiles.entries()).map(([key, fileDetail], i) => (
```

File: actions/knowledge/filehelper.ts
Comment (ryanhopperlowe) [inline actions/knowledge/filehelper.ts:38]: are there going to be many more types of files to import?

File: components/knowledge/FileModal.tsx
Comment (ryanhopperlowe) [inline components/knowledge/FileModal.tsx:31]: this is going to load indefinitely if `runNotionSync` throws an error

File: components/knowledge/FileModal.tsx
Comment (ryanhopperlowe) [inline components/knowledge/FileModal.tsx:47]: Could this be replaced with a Button? `div` with onClick handler will have no accessibility features + `bg-white`-like modifiers don't link to themes. In cases like this it's usually best to use the component library so theming/accessibility is applied by default

File: components/knowledge/FileModal.tsx
Comment (ryanhopperlowe) [inline components/knowledge/FileModal.tsx:60]: same here

File: components/knowledge/Notion.tsx
Comment (ryanhopperlowe) [inline components/knowledge/Notion.tsx:79]: nit: same here. Destructuring is easier to understand
```suggestion
      for (const [fileName, fileDetail] of Array.from(newMap.entries())) {
```

File: components/knowledge/Notion.tsx
Comment (ryanhopperlowe) [inline components/knowledge/Notion.tsx:179]: why use `window.open` here?

File: contexts/edit.tsx
Comment (ryanhopperlowe) [inline contexts/edit.tsx:150]: can probably be removed :)

File: components/knowledge/Notion.tsx
Comment (StrongMonkey) [reply components/knowledge/Notion.tsx:179]: Any good alternative?

File: components/knowledge/FileModal.tsx
Comment (StrongMonkey) [reply components/knowledge/FileModal.tsx:47]: Copied from tailwind template, let me see if I can get button show the same style 👍

File: components/knowledge/FileModal.tsx
Comment (StrongMonkey) [reply components/knowledge/FileModal.tsx:31]: Good call, will fix 👍

File: actions/knowledge/filehelper.ts
Comment (StrongMonkey) [reply actions/knowledge/filehelper.ts:38]: Yes, more types as we add more integration

File: components/knowledge/Notion.tsx
Comment (ryanhopperlowe) [reply components/knowledge/Notion.tsx]: Never seen it used in React before, and wanted to see what the intention here was before I gave an opinion.

Comment (ryanhopperlowe) [reply components/knowledge/Notion.tsx]: Usually for rerouting users use `useRouter` (client + import from `next/navigation`) or `redirect` (server)

But for opening external links in the users default browser I'd go with `<Link isExternal />` imported from `@nextui-org/react`

Not sure if these fit your use case though so feel free to chime in

Comment (StrongMonkey) [reply components/knowledge/Notion.tsx]: Yeah I will change it to use Link. I have no experience though.

Comment (ryanhopperlowe) [reply components/knowledge/Notion.tsx]: Looks Great now!

**Review by ryanhopperlowe — APPROVED** [review]

**Comment by tylerslaton** [comment]
Does this implementation make it so my local machine is a mirror of whatever account I've selected for Notion?

File: components/knowledge/Notion.tsx
Comment (tylerslaton) [inline components/knowledge/Notion.tsx:196]: super nit:

This makes it so that the link doesn't take up a ton of space.
```suggestion
                        <TableCell className="w-[80px]">
```

<details>
<summary>It looks like this</summary>

Screenshot 2024-09-05 at 4 59 31 PM


</details>

File: components/knowledge/Notion.tsx
Comment (tylerslaton) [inline components/knowledge/Notion.tsx:134]: nit and unimportant but:

```suggestion
              <Avatar size="sm" src="notion.svg" alt="Notion Icon" classNames={{base: "p-1.5 bg-white"}}/>
```
<details>

<summary>This makes it so that the logo isn't cut-off.</summary>

Screenshot 2024-09-05 at 5 05 22 PM

**Review by tylerslaton — CHANGES_REQUESTED** [review]
Just ran through this flow and had some thoughts about it but overall this looks good and functions really well! Most of this is just UX questions.

1. Could/should we add an option to settings where I can manage all of this instead of having to go to the edit page
2. It may make sense to add this to the "Add Knowledge" section of a chat interaction. However, that currently doesn't have parity with this either so I'll leave that to your and @cjellick's discretion.
3. When I open choose to add a file through notion and then login I come back and eventually have to push the "sync" button after a successful login. Can we instead make it so that it immediately performs a "first time" sync automatically?
4. Why did we go with syncing the files down locally? Is there anyway that we could do a "just in time" sync where Notion is the source of truth and I pull from it whenever I actually add a file?

**Comment by StrongMonkey** [comment]
@tylerslaton 
> When I open choose to add a file through notion and then login I come back and eventually have to push the "sync" button after a successful login. Can we instead make it so that it immediately performs a "first time" sync automatically?
 It should perform the sync at the first time. I will look at it if it doesn't perform this way.
 
 >Why did we go with syncing the files down locally? Is there anyway that we could do a "just in time" sync where Notion is the source of truth and I pull from it whenever I actually add a file?
 Yeah, I kind of went back and forth on this. With onedrive I am going to change this so we only sync down files when you add it into assistant, because in onedrive you can't select the file/folders when you do oauth. For notion it is less of an issue because you can select the pages when doing oauth and that will limit the files it syncs initially. The benefit of syncing down all the files is that later when you add files through assistants, it is relatively fast but yes the initial sync will be slow.

**Comment by StrongMonkey** [comment]
>Does this implementation make it so my local machine is a mirror of whatever account I've selected for Notion?

Yes, files that you select through notion will eventually match files in your local laptop.

**Comment by StrongMonkey** [comment]
I think I am going to just change the import logic to sync on demand for now. This means that we will only start syncing when the file has been used by assistants.

**Comment by StrongMonkey** [comment]
>Could/should we add an option to settings where I can manage all of this instead of having to go to the edit page
It may make sense to add this to the "Add Knowledge" section of a chat interaction. However, that currently doesn't have parity with this either so I'll leave that to your and @cjellick's discretion.

Good point, i will fix `Add knowledge` to bring parity. For setting page I am not sure, I leans towards not adding it in setting page for now for simplicity of the feature but open to opinions for @cjellick.

**Comment by StrongMonkey** [comment]
@tylerslaton Most of your comments shoud be addressed.
>Could/should we add an option to settings where I can manage all of this instead of having to go to the edit page

I think we are just going to punt on this until we have more UX feedback from craig and will

>It may make sense to add this to the "Add Knowledge" section of a chat interaction. However, that currently doesn't have parity with this either so I'll leave that to your and @cjellick's discretion.

Per standup we are not going to do this

>When I open choose to add a file through notion and then login I come back and eventually have to push the "sync" button after a successful login. Can we instead make it so that it immediately performs a "first time" sync automatically?

This should run the sync for first time, if it is still happening to you we can sync, doesn't look like the case for me


>Why did we go with syncing the files down locally? Is there anyway that we could do a "just in time" sync where Notion is the source of truth and I pull from it whenever I actually add a file?

So I changed this to be synced on demand. The initial sync will only fetch metadata.

**Review by tylerslaton — COMMENTED** [review]
Things are looking good testing locally. Going to do a second pass on the code in the morning and likely approve around then. Thanks for all the addressing of my feedback I really appreciate it. The feature is looking awesome, I love how it all flows.

File: actions/knowledge/filehelper.ts
Comment (tylerslaton) [inline actions/knowledge/filehelper.ts:40]: Why are we calling path.dirname twice? What is the path we're retrieving? Whatever the answer, it could be a good comment.

File: actions/knowledge/filehelper.ts
Comment (ryanhopperlowe) [inline actions/knowledge/filehelper.ts:35]: nit: it looks like the return type of this method can be inferred automatically. It's usually a good idea in TS to use inferred return types so that the compiler can catch errors at compile time if anything changes inside of the function.

File: server/app.mjs
Comment (tylerslaton) [inline server/app.mjs:62]: nit: mind putting a comment for the actual time here?

**Review by tylerslaton — DISMISSED** [review]
No more blockers for me, just minor nits. Lots of code here so I'd like more eyes on this.

File: actions/knowledge/util.ts
Comment (ryanhopperlowe) [inline actions/knowledge/util.ts:1]: one more thing that would be helpful is to move this type to the `@/model` directory?

**Comment by cjellick** [comment]
These are my notes on the feature thus far...its raw and unprocessed and i dont have solutins, just wanted to get it out there

1. the plus’s functionality is not obvious
2. the feedback while its pulling the metadata is not good
3. the paths could be better. maybe by just making them relative?
4. once i added the files, it was all my files plus the ones in that dir. that was bad
    1. this is the probably the worst thing.
    2. it also should probably have multi-select via range.
    3. i think once i click add, it should be a page dedicated to just those items. or maybe just sorting better would be good enough. maybe a "clear" button to get rid of everything in th emodal
5. check is very sluggish. dont like the animation
6. this probably doesnt cut it for future changes to the directory (to sync a diff of files)
7. error messaging and handling is terrible. outputs all the logs as errors

File: actions/knowledge/util.ts
Comment (StrongMonkey) [reply actions/knowledge/util.ts:5]: Yeah sounds good to me, I will change that

**Comment by StrongMonkey** [comment]
@tylerslaton @ryanhopperlowe should have addressed all your comments.

**Comment by ryanhopperlowe** [comment]
Everything looks good, except I think you missed 1 thing. Once that's fixed I'll shoot the approval :)

File: contexts/edit.tsx
Comment (StrongMonkey) [reply contexts/edit.tsx:61]: @ryanhopperlowe I DRYed the code bit, but it is not possible to move all the data convertions because setDroppedFiles is used in multiple places where update and delete happens with different implementation.

**Review by ryanhopperlowe — APPROVED** [review]

**Review by tylerslaton — DISMISSED** [review]
My comments have all been addressed. Thanks!

**Comment by tylerslaton** [comment]
Oh and definitely be sure to fix the merge conflicts 👍🏻

File: components/knowledge/OneDrive.tsx
Comment (ryanhopperlowe) [inline components/knowledge/OneDrive.tsx:144]: nit: you can get most of this logic out of the box with the `useAsync` hook that I created

**Review by ryanhopperlowe — DISMISSED** [review]

**Review by ryanhopperlowe — APPROVED** [review]

**Review by tylerslaton — APPROVED** [review]

**Pull request merged**
